proplet/mqtt.go

```diff
@@ -0,0 +1,185 @@
+	})
+
+	opts.SetReconnectingHandler(func(client mqtt.Client, options *mqtt.ClientOptions) {
+		fmt.Println("MQTT reconnecting...")
+	})
+
+	client := mqtt.NewClient(opts)
+	token := client.Connect()
+	if token.Wait() && token.Error() != nil {
+		return nil, fmt.Errorf("failed to connect to MQTT broker '%s': %w", config.BrokerURL, pkgerrors.ErrMQTTConnectionFailed)
+	}
+
+	fmt.Println("MQTT client connected successfully.")
+
+	// Publish discovery message
+	if err := PublishDiscovery(client, config); err != nil {
+		return nil, fmt.Errorf("failed to publish discovery message: %w", err)
+	}
+
+	// Start liveliness updates
+	go startLivelinessUpdates(client, config)
+
+	return client, nil
+}
+
+// validateConfig ensures that the provided configuration is valid and complete.
+func validateConfig(config *Config) error {
+	if config == nil {
+		return fmt.Errorf("config is nil: %w", pkgerrors.ErrConfigValidation)
+	}
+	if config.ChannelID == "" {
+		return fmt.Errorf("ChannelID is missing: %w", pkgerrors.ErrMissingChannelID)
+	}
+	if config.PropletID == "" {
+		return fmt.Errorf("PropletID is missing: %w", pkgerrors.ErrMissingPropletID)
+	}
+	if config.BrokerURL == "" {
+		return fmt.Errorf("BrokerURL is missing: %w", pkgerrors.ErrMissingValue)
+	}
+	if _, err := url.ParseRequestURI(config.BrokerURL); err != nil {
+		return fmt.Errorf("invalid broker URL '%s': %w", config.BrokerURL, pkgerrors.ErrInvalidValue)
+	}
+	return nil
+}
+
+// publishDiscovery sends an initial "create" message to announce the Proplet's existence.
+func PublishDiscovery(client mqtt.Client, config *Config) error {
+	topic := fmt.Sprintf("channels/%s/messages/control/proplet/create", config.ChannelID)
+	payload := fmt.Sprintf(`{"proplet_id":"%s","chan_id":"%s"}`, config.PropletID, config.ChannelID)
+	token := client.Publish(topic, 0, false, payload)
+	token.Wait()
+	if token.Error() != nil {
+		return fmt.Errorf("failed to publish discovery message: %w", token.Error())
+	}
+	fmt.Printf("Published discovery message to topic '%s'\n", topic)
+	return nil
+}
+
+// startLivelinessUpdates sends periodic "alive" messages to the MQTT broker.
+func startLivelinessUpdates(client mqtt.Client, config *Config) {
+	for {
+		topic := fmt.Sprintf("channels/%s/messages/control/proplet/alive", config.ChannelID)
+		payload := fmt.Sprintf(`{"status":"alive","proplet_id":"%s","chan_id":"%s"}`, config.PropletID, config.ChannelID)
+		token := client.Publish(topic, 0, false, payload)
+		token.Wait()
+		if token.Error() != nil {
+			fmt.Printf("Failed to publish liveliness: %v\n", token.Error())
+		}
+		time.Sleep(10 * time.Second)
+	}
+}
+
+// SubscribeToTopics subscribes to relevant MQTT topics for Manager and registry interaction.
+func SubscribeToManagerTopics(client mqtt.Client, config *Config, startHandler, stopHandler, registryHandler mqtt.MessageHandler) error {
+	// Subscribe to the start command topic
+	startTopic := fmt.Sprintf("channels/%s/messages/control/manager/start", config.ChannelID)
```

Comment:
define topics at the top and use variable substitution when you want to use them